### PR TITLE
Bypass cudaStreamLegacy related bug that exists in early version of NCCL

### PR DESCRIPTION
**Pull request by TarzanZhao**
There is a bug related to cudaStreamLegacy (default stream) in early version of NCCL. This PR solves it. it uses cupy's solution that bypasses this bug with 0 pointer (in below picture). 

image

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h [hunk 1]
```diff
@@ -70,29 +69,35 @@ ncclDataType_t ToNcclDataType(PrimitiveType element_type);
 
 int SizeOfType(ncclDataType_t element_type);
 
-StatusOr< std::shared_ptr< std::vector<ncclComm_t> > > NcclInitCommunicator(std::vector<int> devices_vec);
+class NcclCommStorage {
```
Comment (zhuohan123) [inline tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h:72]: Nit: `Storage` is a general name that does not indicate what this part is doing, can you rename it to something more concrete? Maybe `NcclCommAndStreamVector`?

**Review by zhuohan123 — APPROVED** [review]
LGTM.

You can also format c++ files following this guide: https://github.com/tensorflow/tensorflow/blob/master/CONTRIBUTING.md#c-coding-style

File: tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h (hunk 1 above)
Comment (TarzanZhao) [reply tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h:72]: OK

Comment (TarzanZhao) [reply tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.h:72]: Well, I might add other information(e.g. cuda event) in this storage later, so I use the word `storage` here.

File: tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc [hunk 2]
```diff
@@ -246,15 +264,16 @@ StatusOr<int> NcclGetVersion() {
 #endif  // XLA_ENABLE_XCCL
 }
 
-StatusOr< std::shared_ptr< std::vector<ncclComm_t> > > NcclCreateCommunicators(int world_size, 
-                                                                               std::vector<int> devices_global_rank, 
-                                                                               std::vector<int> devices_ids, 
-                                                                               std::vector<char> nccl_uid_vec) {
+StatusOr< std::shared_ptr<NcclCommStorage> > NcclCreateCommunicators(int world_size,
```
Comment (merrymercy) [inline tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc:267]: Code style:
```
StatusOr<std::shared_ptr<NcclCommStorage>> NcclCreateCommunicators(int world_size, 
```

File: tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc [hunk 3]
```diff
@@ -263,7 +282,21 @@ StatusOr< std::shared_ptr< std::vector<ncclComm_t> > > NcclCreateCommunicators(i
     XLA_CUDA_RETURN_IF_ERROR(ncclCommInitRank(comms.data()+i, world_size, nccl_uid, devices_global_rank[i]));
   }
   XLA_CUDA_RETURN_IF_ERROR(ncclGroupEnd());
-  return std::make_shared< std::vector<ncclComm_t> >(std::move(comms));
+
+  std::vector<cudaStream_t> streams;
+  streams.resize(n_devices);
+  for (int i=0; i<n_devices; i++) {
+    cudaSetDevice(devices_ids[i]);
+    if (nccl_use_multistream)
```
Comment (merrymercy) [inline tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc:290]: code style:
always use `{}` to wrap if-else block even it only has one line

File: tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc (hunk 2 above)
Comment (TarzanZhao) [reply tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc:267]: OK, I will create another PR to delete cupy nccl later. I will modify what you mentioned here along with that PR.

File: tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc (hunk 3 above)
Comment (TarzanZhao) [reply tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc:290]: OK, get it.

**Comment by merrymercy** [comment]
Also, some of these header files are not used. https://github.com/alpa-projects/tensorflow-alpa/blob/c6da162d80954c00fc9ca3d0848e5300ffbc980a/tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc#L24-L33
Please delete them in later PRs.

**Comment by TarzanZhao** [comment]
> Also, some of these header files are not used.
> 
> https://github.com/alpa-projects/tensorflow-alpa/blob/c6da162d80954c00fc9ca3d0848e5300ffbc980a/tensorflow/compiler/xla/service/gpu/alpa_nccl_wrapper.cc#L24-L33
> 
> 
> Please delete them in later PRs.

OK

**Pull request merged**